In [1]:
%matplotlib inline


In [3]:
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split


In [6]:
import  seaborn as sns
from scipy.stats import zscore

In [7]:
from sklearn.neighbors import KNeighborsClassifier

In [8]:
NNH=KNeighborsClassifier(n_neighbors=5,weights='distance')

In [9]:
bc_df=pd.read_csv("C:/Users/Narasimha Rao/OneDrive/Desktop/wisc.csv")

In [10]:
bc_df.shape

(569, 32)

In [11]:
bc_df.dtypes

id                     int64
diagnosis             object
radius_mean          float64
texture_mean         float64
perimeter_mean       float64
area_mean            float64
smoothness_mean      float64
compactness_mean     float64
concavity_mean       float64
points_mean          float64
symmetry_mean        float64
dimension_mean       float64
radius_se            float64
texture_se           float64
perimeter_se         float64
area_se              float64
smoothness_se        float64
compactness_se       float64
concavity_se         float64
points_se            float64
symmetry_se          float64
dimension_se         float64
radius_worst         float64
texture_worst        float64
perimeter_worst      float64
area_worst           float64
smoothness_worst     float64
compactness_worst    float64
concavity_worst      float64
points_worst         float64
symmetry_worst       float64
dimension_worst      float64
dtype: object

object type is string type..as none of the algorithm works on a strings..so we have to chang type of dignosis column

In [12]:
bc_df["diagnosis"]=bc_df.diagnosis.astype("category")

In [13]:
bc_df.dtypes

id                      int64
diagnosis            category
radius_mean           float64
texture_mean          float64
perimeter_mean        float64
area_mean             float64
smoothness_mean       float64
compactness_mean      float64
concavity_mean        float64
points_mean           float64
symmetry_mean         float64
dimension_mean        float64
radius_se             float64
texture_se            float64
perimeter_se          float64
area_se               float64
smoothness_se         float64
compactness_se        float64
concavity_se          float64
points_se             float64
symmetry_se           float64
dimension_se          float64
radius_worst          float64
texture_worst         float64
perimeter_worst       float64
area_worst            float64
smoothness_worst      float64
compactness_worst     float64
concavity_worst       float64
points_worst          float64
symmetry_worst        float64
dimension_worst       float64
dtype: object

category is stored as numbers internally

for xample employee id..we can not use empid in any mathematical calculations..they are only nemerical values

In [14]:
bc_df.describe().transpose()

count          mean           std          min  \
id                 569.0  3.037183e+07  1.250206e+08  8670.000000   
radius_mean        569.0  1.412729e+01  3.524049e+00     6.981000   
texture_mean       569.0  1.928965e+01  4.301036e+00     9.710000   
perimeter_mean     569.0  9.196903e+01  2.429898e+01    43.790000   
area_mean          569.0  6.548891e+02  3.519141e+02   143.500000   
smoothness_mean    569.0  9.636028e-02  1.406413e-02     0.052630   
compactness_mean   569.0  1.043410e-01  5.281276e-02     0.019380   
concavity_mean     569.0  8.879932e-02  7.971981e-02     0.000000   
points_mean        569.0  4.891915e-02  3.880284e-02     0.000000   
symmetry_mean      569.0  1.811619e-01  2.741428e-02     0.106000   
dimension_mean     569.0  6.279761e-02  7.060363e-03     0.049960   
radius_se          569.0  4.051721e-01  2.773127e-01     0.111500   
texture_se         569.0  1.216853e+00  5.516484e-01     0.360200   
perimeter_se       569.0  2.866059e+00  2.021855e+00     0.757000   
area_se            569.0  4.033708e+01  4.549101e+01     6.802000   
smoothness_se      569.0  7.040979e-03  3.002518e-03     0.001713   
compactness_se     569.0  2.547814e-02  1.790818e-02     0.002252   
concavity_se       569.0  3.189372e-02  3.018606e-02     0.000000   
points_se          569.0  1.179614e-02  6.170285e-03     0.000000   
symmetry_se        569.0  2.054230e-02  8.266372e-03     0.007882   
dimension_se       569.0  3.794904e-03  2.646071e-03     0.000895   
radius_worst       569.0  1.626919e+01  4.833242e+00     7.930000   
texture_worst      569.0  2.567722e+01  6.146258e+00    12.020000   
perimeter_worst    569.0  1.072612e+02  3.360254e+01    50.410000   
area_worst         569.0  8.805831e+02  5.693570e+02   185.200000   
smoothness_worst   569.0  1.323686e-01  2.283243e-02     0.071170   
compactness_worst  569.0  2.542650e-01  1.573365e-01     0.027290   
concavity_worst    569.0  2.721885e-01  2.086243e-01     0.000000   
points_worst       569.0  1.146062e-01  6.573234e-02     0.000000   
symmetry_worst     569.0  2.900756e-01  6.186747e-02     0.156500   
dimension_worst    569.0  8.394582e-02  1.806127e-02     0.055040   

                             25%            50%           75%           max  
id                 869218.000000  906024.000000  8.813129e+06  9.113205e+08  
radius_mean            11.700000      13.370000  1.578000e+01  2.811000e+01  
texture_mean           16.170000      18.840000  2.180000e+01  3.928000e+01  
perimeter_mean         75.170000      86.240000  1.041000e+02  1.885000e+02  
area_mean             420.300000     551.100000  7.827000e+02  2.501000e+03  
smoothness_mean         0.086370       0.095870  1.053000e-01  1.634000e-01  
compactness_mean        0.064920       0.092630  1.304000e-01  3.454000e-01  
concavity_mean          0.029560       0.061540  1.307000e-01  4.268000e-01  
points_mean             0.020310       0.033500  7.400000e-02  2.012000e-01  
symmetry_mean           0.161900       0.179200  1.957000e-01  3.040000e-01  
dimension_mean          0.057700       0.061540  6.612000e-02  9.744000e-02  
radius_se               0.232400       0.324200  4.789000e-01  2.873000e+00  
texture_se              0.833900       1.108000  1.474000e+00  4.885000e+00  
perimeter_se            1.606000       2.287000  3.357000e+00  2.198000e+01  
area_se                17.850000      24.530000  4.519000e+01  5.422000e+02  
smoothness_se           0.005169       0.006380  8.146000e-03  3.113000e-02  
compactness_se          0.013080       0.020450  3.245000e-02  1.354000e-01  
concavity_se            0.015090       0.025890  4.205000e-02  3.960000e-01  
points_se               0.007638       0.010930  1.471000e-02  5.279000e-02  
symmetry_se             0.015160       0.018730  2.348000e-02  7.895000e-02  
dimension_se            0.002248       0.003187  4.558000e-03  2.984000e-02  
radius_worst           13.010000      14.970000  1.879000e+01  3.604000e+01  
texture_worst   

look at in to the data before model building..
is  there any relationship between dignosis and other features..
in knn we dont have any model..
in dignosis after putting the filter we have B and M..WHAT are the short forms meaning
in some features we have 0..is it missing at random or it is a value

after sorting id of patients we observr some jumps ..where is the remaining data..
we have to ask about data..if there is a valid reason we move on


in radius_mean...if we draw a grapg for distibution..we have min max..mean,,q1,q2,q3
the tails are distance between q1 and min and max and q3

there are outliers because of these tails are there

In [24]:
bc_df.groupby(["diagnosis"]).count()

id  radius_mean  texture_mean  perimeter_mean  area_mean  \
diagnosis                                                              
B          357          357           357             357        357   
M          212          212           212             212        212   

           smoothness_mean  compactness_mean  concavity_mean  points_mean  \
diagnosis                                                                   
B                      357               357             357          357   
M                      212               212             212          212   

           symmetry_mean  ...  radius_worst  texture_worst  perimeter_worst  \
diagnosis                 ...                                                 
B                    357  ...           357            357              357   
M                    212  ...           212            212              212   

           area_worst  smoothness_worst  compactness_worst  concavity_worst  \
diagnosis                                                                     
B                 357               357                357              357   
M                 212               212                212              212   

           points_worst  symmetry_worst  dimension_worst  
diagnosis                                                 
B                   357             357              357  
M                   212             212              212  

[2 rows x 31 columns]

drop thr first column we are not using this in modeling


In [25]:
bc_feature_df=bc_df.drop(labels="id",axis=1)

we use group by for diagnosis...it separates the data based on the category..
in our dataset..we have B=357 and M=212

our model best identifies B cases than M cases...
is our model can identify all the cases using 357 benain  data points
and 212 cases by using 212 malignant case

our models can get malignant as malignant
aditional sampling or reduce the benain
we didnt reduce the we have add the data


the model that we crete model on large data it is better modl




In [32]:
bc_feature_df

diagnosis  radius_mean  texture_mean  perimeter_mean  area_mean  \
0           B        12.32         12.39           78.85      464.1   
1           B        10.60         18.95           69.28      346.4   
2           B        11.04         16.83           70.92      373.2   
3           B        11.28         13.39           73.00      384.8   
4           B        15.19         13.21           97.65      711.8   
..        ...          ...           ...             ...        ...   
564         B        13.17         18.22           84.28      537.3   
565         B        10.26         14.71           66.20      321.6   
566         M        15.28         22.41           98.92      710.6   
567         B        14.53         13.98           93.86      644.2   
568         M        21.37         15.10          141.30     1386.0   

     smoothness_mean  compactness_mean  concavity_mean  points_mean  \
0            0.10280           0.06981         0.03987      0.03700   
1            0.09688           0.11470         0.06387      0.02642   
2            0.10770           0.07804         0.03046      0.02480   
3            0.11640           0.11360         0.04635      0.04796   
4            0.07963           0.06934         0.03393      0.02657   
..               ...               ...             ...          ...   
564          0.07466           0.05994         0.04859      0.02870   
565          0.09882           0.09159         0.03581      0.02037   
566          0.09057           0.10520         0.05375      0.03263   
567          0.10990           0.09242         0.06895      0.06495   
568          0.10010           0.15150         0.19320      0.12550   

     symmetry_mean  ...  radius_worst  texture_worst  perimeter_worst  \
0           0.1959  ...         13.50          15.64            86.97   
1           0.1922  ...         11.88          22.94            78.28   
2           0.1714  ...         12.41          26.44            79.93   
3           0.1771  ...         11.92          15.77            76.53   
4           0.1721  ...         16.20          15.73           104.50   
..             ...  ...           ...            ...              ...   
564         0.1454  ...         14.90          23.89            95.10   
565         0.1633  ...         10.88          19.48            70.89   
566         0.1727  ...         17.80          28.03           113.80   
567         0.1650  ...         15.80          16.93           103.10   
568         0.1973  ...         22.69          21.84           152.10   

     area_worst  smoothness_worst  compactness_worst  concavity_worst  \
0         549.1            0.1385             0.1266          0.12420   
1         424.8            0.1213             0.2515          0.19160   
2         471.4            0.1369             0.1482          0.10670   
3         434.0            0.1367             0.1822          0.08669   
4         819.1            0.1126             0.1737          0.13620   
..          ...               ...                ...              ...   
564       687.6            0.1282             0.1965          0.18760   
565       357.1            0.1360             0.1636          0.07162   
566       973.1            0.1301             0.3299          0.36300   
567       749.9            0.1347             0.1478          0.13730   
568      1535.0            0.1192             0.2840          0.40240   

     points_worst  symmetry_worst  dimension_worst  
0         0.09391          0.2827          0.06771  
1         0.07926          0.2940          0.07587  
2         0.07431          0.2998          0.07881  
3         0.08611          0.2102          0.06784  
4         0.08178          0.2487          0.06766  
..            ...             ...              ...  
564       0.10450          0.2235          0.06925  
565       0.04074          0.2434          0.08488  
566       0.12260          0.3175          0.09772  
567       0.10

In [33]:
bc_feature_df1=bc_feature_df.drop(labels="diagnosis",axis=1)

In [34]:
bc_feature_df1


radius_mean  texture_mean  perimeter_mean  area_mean  smoothness_mean  \
0          12.32         12.39           78.85      464.1          0.10280   
1          10.60         18.95           69.28      346.4          0.09688   
2          11.04         16.83           70.92      373.2          0.10770   
3          11.28         13.39           73.00      384.8          0.11640   
4          15.19         13.21           97.65      711.8          0.07963   
..           ...           ...             ...        ...              ...   
564        13.17         18.22           84.28      537.3          0.07466   
565        10.26         14.71           66.20      321.6          0.09882   
566        15.28         22.41           98.92      710.6          0.09057   
567        14.53         13.98           93.86      644.2          0.10990   
568        21.37         15.10          141.30     1386.0          0.10010   

     compactness_mean  concavity_mean  points_mean  symmetry_mean  \
0             0.06981         0.03987      0.03700         0.1959   
1             0.11470         0.06387      0.02642         0.1922   
2             0.07804         0.03046      0.02480         0.1714   
3             0.11360         0.04635      0.04796         0.1771   
4             0.06934         0.03393      0.02657         0.1721   
..                ...             ...          ...            ...   
564           0.05994         0.04859      0.02870         0.1454   
565           0.09159         0.03581      0.02037         0.1633   
566           0.10520         0.05375      0.03263         0.1727   
567           0.09242         0.06895      0.06495         0.1650   
568           0.15150         0.19320      0.12550         0.1973   

     dimension_mean  ...  radius_worst  texture_worst  perimeter_worst  \
0           0.05955  ...         13.50          15.64            86.97   
1           0.06491  ...         11.88          22.94            78.28   
2           0.06340  ...         12.41          26.44            79.93   
3           0.06072  ...         11.92          15.77            76.53   
4           0.05544  ...         16.20          15.73           104.50   
..              ...  ...           ...            ...              ...   
564         0.05549  ...         14.90          23.89            95.10   
565         0.07005  ...         10.88          19.48            70.89   
566         0.06317  ...         17.80          28.03           113.80   
567         0.06121  ...         15.80          16.93           103.10   
568         0.06183  ...         22.69          21.84           152.10   

     area_worst  smoothness_worst  compactness_worst  concavity_worst  \
0         549.1            0.1385             0.1266          0.12420   
1         424.8            0.1213             0.2515          0.19160   
2         471.4            0.1369             0.1482          0.10670   
3         434.0            0.1367             0.1822          0.08669   
4         819.1            0.1126             0.1737          0.13620   
..          ...               ...                ...              ...   
564       687.6            0.1282             0.1965          0.18760   
565       357.1            0.1360             0.1636          0.07162   
566       973.1            0.1301             0.3299          0.36300   
567       749.9            0.1347             0.1478          0.13730   
568      1535.0            0.1192             0.2840          0.40240   

     points_worst  symmetry_worst  dimension_worst  
0         0.09391          0.2827          0.06771  
1         0.07926          0.2940          0.07587  
2         0.07431          0.2998          0.07881  
3         0.08611          0.2102          0.06784  
4         0.08178          0.2487          0.06766  
..            ...             ...              ...  
564       0.10450          0.2235          0.06925  
565       0.04074          0.2434          0.0848

In [37]:
bc_feature_df1_z=bc_feature_df1.apply(zscore)

In [38]:
bc_feature_df1_z.describe()

radius_mean  texture_mean  perimeter_mean     area_mean  \
count  5.690000e+02  5.690000e+02    5.690000e+02  5.690000e+02   
mean  -1.287781e-16 -7.373520e-16   -1.303390e-16 -2.239958e-16   
std    1.000880e+00  1.000880e+00    1.000880e+00  1.000880e+00   
min   -2.029648e+00 -2.229249e+00   -1.984504e+00 -1.454443e+00   
25%   -6.893853e-01 -7.259631e-01   -6.919555e-01 -6.671955e-01   
50%   -2.150816e-01 -1.046362e-01   -2.359800e-01 -2.951869e-01   
75%    4.693926e-01  5.841756e-01    4.996769e-01  3.635073e-01   
max    3.971288e+00  4.651889e+00    3.976130e+00  5.250529e+00   

       smoothness_mean  compactness_mean  concavity_mean   points_mean  \
count     5.690000e+02      5.690000e+02    5.690000e+02  5.690000e+02   
mean      1.660457e-16      1.882891e-16    5.112099e-17 -4.760886e-17   
std       1.000880e+00      1.000880e+00    1.000880e+00  1.000880e+00   
min      -3.112085e+00     -1.610136e+00   -1.114873e+00 -1.261820e+00   
25%      -7.109628e-01     -7.470860e-01   -7.437479e-01 -7.379438e-01   
50%      -3.489108e-02     -2.219405e-01   -3.422399e-01 -3.977212e-01   
75%       6.361990e-01      4.938569e-01    5.260619e-01  6.469351e-01   
max       4.770911e+00      4.568425e+00    4.243589e+00  3.927930e+00   

       symmetry_mean  dimension_mean  ...  radius_worst  texture_worst  \
count   5.690000e+02    5.690000e+02  ...  5.690000e+02   5.690000e+02   
mean    1.465338e-16   -1.517240e-15  ... -8.507157e-17   6.634022e-18   
std     1.000880e+00    1.000880e+00  ...  1.000880e+00   1.000880e+00   
min    -2.744117e+00   -1.819865e+00  ... -1.726901e+00  -2.223994e+00   
25%    -7.032397e-01   -7.226392e-01  ... -6.749213e-01  -7.486293e-01   
50%    -7.162650e-02   -1.782793e-01  ... -2.690395e-01  -4.351564e-02   
75%     5.307792e-01    4.709834e-01  ...  5.220158e-01   6.583411e-01   
max     4.484751e+00    4.910919e+00  ...  4.094189e+00   3.885905e+00   

       perimeter_worst    area_worst  smoothness_worst  compactness_worst  \
count     5.690000e+02  5.690000e+02      5.690000e+02       5.690000e+02   
mean     -3.570665e-16 -1.888745e-16     -2.234104e-16       7.999850e-18   
std       1.000880e+00  1.000880e+00      1.000880e+00       1.000880e+00   
min      -1.693361e+00 -1.222423e+00     -2.682695e+00      -1.443878e+00   
25%      -6.895783e-01 -6.421359e-01     -6.912304e-01      -6.810833e-01   
50%      -2.859802e-01 -3.411812e-01     -4.684277e-02      -2.695009e-01   
75%       5.402790e-01  3.575891e-01      5.975448e-01       5.396688e-01   
max       4.287337e+00  5.930172e+00      3.955374e+00       5.112877e+00   

       concavity_worst  points_worst  symmetry_worst  dimension_worst  
count     5.690000e+02  5.690000e+02    5.690000e+02     5.690000e+02  
mean      9.033976e-17  2.224348e-16    2.734583e-16     1.851672e-16  
std       1.000880e+00  1.000880e+00    1.000880e+00     1.000880e+00  
min      -1.305831e+00 -1.745063e+00   -2.160960e+00    -1.601839e+00  
25%      -7.565142e-01 -7.563999e-01   -6.418637e-01    -6.919118e-01  
50%      -2.182321e-01 -2.234689e-01   -1.274095e-01    -2.164441e-01  
75%       5.311411e-01  7.125100e-01    4.501382e-01     4.507624e-01  
max       4.700669e+00  2.685877e+00    6.046041e+00     6.846856e+00  

[8 rows x 30 columns]

z score is equl to =x-mean/standard deviation....

knn is used for distance calculations..so all are in same scale thats why we used scaling

In [39]:
bc_labels=bc_df["diagnosis"]

In [41]:
x=np.array(bc_feature_df1_z)

In [42]:
x.shape

(569, 30)

In [43]:
y=np.array(bc_labels)

In [44]:
y.shape

(569,)

In [45]:
x_train,x_test,y_train,y_test=train_test_split(x,y,test_size=0.30,random_state=1)pre

In [46]:
NNH.fit(x_train,y_train)

KNeighborsClassifier(weights='distance')

In [47]:
predicted_labels=NNH.predict(x_test)

In [48]:
NNH.score(x_test,y_test)

0.9649122807017544

In [49]:
from  sklearn import metrics
print(metrics.confusion_matrix(y_test,predicted_labels))

[[112   2]
 [  4  53]]
